website/docs/gator.md

```diff
@@ -427,6 +427,80 @@ However, not including the `namespace` definition in the call to `gator expand`
 error expanding resources: error expanding resource nginx-deployment: failed to mutate resultant resource nginx-deployment-pod: matching for mutator Assign.mutations.gatekeeper.sh /always-pull-image failed for  Pod my-ns nginx-deployment-pod: failed to run Match criteria: namespace selector for namespace-scoped object but missing Namespace
 ```
 
+## The `gator sync verify` subcommand
+
+Certain templates require [replicating data](sync.md) into OPA to enable correct evaluation. These templates can use the annotation `metadata.gatekeeper.sh/requires-sync-data` to indicate which resources need to be synced. The annotation contains a json object representing a list of requirements, each of which contains a list of one or more equivalence sets. Each of these equivalence sets has `groups`, `versions`, and `kinds` fields; any group-version-kind combination within an equivalence set within a requirement should be considered sufficient to satisfy that requirement. For example (comments added for clarity):
+```
+[
+  [ // Requirement 1
+    { // Equivalence set 1
+      "groups": ["group1", group2"]
+      "versions": ["version1", "version2", "version3"]
+      "kinds": ["kind1", "kind2"]
+    },
+    { // Equivalence Set 2
+      "groups": ["group3", group4"]
+      "versions": ["version3", "version4"]
+      "kinds": ["kind3", "kind4"]
+    }
```

Comment:
Aren't these actually one Equivalence set?

To take the example from https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/general/uniqueingresshost/template.yaml:

```
      "[
        [
          {
            "groups": ["extensions"],
            "versions": ["v1beta1"],
            "kinds": ["Ingress"]
          },
          {
            "groups": ["networking.k8s.io"],
            "versions": ["v1beta1", "v1"],
            "kinds": ["Ingress"]
          }
        ]
      ]"
```

All the following GVKs are equivalent:
1. extensions/v1beta1:Ingress
2. networking.k8s.io/v1beta1:Ingress
3. networking.k8s.io/v1:Ingress

They are thus the set of things which are equivalent and should be a single "equivalence set".

Am I misunderstanding this?

Reply:
Good question, the way I've been using it (at least in the [code](https://github.com/open-policy-agent/gatekeeper/blob/a7e3b7c947bcf2833d4cf424ad887c387b4da14f/pkg/cachemanager/parser/syncannotationreader.go#L79)) is that an ES is the specific unit, and multiple of them can be composed into a requirement, the idea being that the unit has a particular syntax (group**s** version**s** kind**s**) and thus needs its own type in code/is its own JSON object in input

Reply:
OK just want to clarify... In the above example, are the [three representations](https://github.com/open-policy-agent/gatekeeper/pull/3098#discussion_r1369368083) all entries in the map of GVKs called `GVKEquivalenceSet` here in the code?

https://github.com/open-policy-agent/gatekeeper/blob/3a83700f287e73376ba264db1cc75d14dd0da8f8/pkg/cachemanager/parser/syncannotationreader.go#L15-L21

If that _isn't_ true, that is a bug.

> at the end of the day I think it still is clearest to be able to call each of the 2 objects within one requirement their own ES, even if they get parsed into a single ES under the hood...thoughts?

I feel the opposite.  IMHO, the terminology should maximize meaning and be extremely true to the intention.

Any set of things that are equivalent is an... `EquivalenceSet`.  If we _don't_ say that, we're saying that an EquivalenceSet is more about how things are written than about what they are.  That's surely confusing.

If we need a name for one of those single JSON structs, let's just make one that is unambiguous: `GVKClause`.  I don't feel strongly about the exact name, I just want us to use our essential terminology (`EquivalenceSet`) wisely.

FWIW, users don't have to specify a type when they make a "GVK Clause", so the naming isn't all that important.  

-------

OH, a further example.  Imagine that Ingress didn't have a `v1` in `networking.k8s.io`.  We'd then write it as:

```
          {
            "groups": ["networking.k8s.io", "extensions"],
            "versions": ["v1beta1"],
            "kinds": ["Ingress"]
          }
```

What was two equivalenceSets became 1.

Another way to think of it is that we could take our current 2 clauses and write them as 3:

```
      "[
        [
          {
            "groups": ["extensions"],
            "versions": ["v1beta1"],
            "kinds": ["Ingress"]
          },
          {
            "groups": ["networking.k8s.io"],
            "versions": ["v1beta1"],
            "kinds": ["Ingress"]
          },
          {
            "groups": ["networking.k8s.io"],
            "versions": ["v1"],
            "kinds": ["Ingress"]
          }
        ]
      ]"
```

Do we now have 3 EQs?  Clearly we still have one set of GVKs which are equivalent to each other.

Reply:
I like clause! Will switch